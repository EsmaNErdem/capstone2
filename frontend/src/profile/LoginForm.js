import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Formik, Field, Form, ErrorMessage } from 'formik';
import { Typography } from '@mui/material';
import Alert from "../utilities/Alert";
import Loading from "../utilities/Loading";
import * as Yup from 'yup';
import "./FormContainer.css"
import { faL } from '@fortawesome/free-solid-svg-icons';

/**
 * LoginForm Component
 * Displays form and renders controlled components functionality
 * Provides form validation with formik
 * 
 * A form component for user login. Handles user input for username, password.
 * Displays form errors if registration fails.
 * On submission:
 * - calls login function prop
 * - redirects to / route
 *
 * Routes ==> LoginForm ==> Alert 
 */
const LoginForm = ({ login }) => {
    console.debug("LoginForm");

    const navigate = useNavigate();
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(false)

    /**
     * Handles form submission. Calls the login function with the form data.
     * Redirects to the home page upon successful registration, or displays form errors.
     */
    const handleSubmit = async (values, { setSubmitting }) => {
        try {
            setLoading(true);
            const result = await login(values);
    
            if (result.success) {
                navigate("/");
                setLoading(false)
            } else {
                setError(result.error);
            }
        } catch (error) {
            setError("An error occurred during login. Please try again...");
            setLoading(false)
        } finally {
            setSubmitting(false);
        }
    };

    return (
        <div>
            <Typography variant="h2" sx={{  marginBottom: 2 }}>Login</Typography>
            {error ? <Alert type="danger" messages={[error]} />: null}
            <Formik
                initialValues={{ username: '', password: ''}}
                validationSchema={Yup.object({
                    username: Yup.string()
                    .max(15, 'Must be 15 characters or less')
                    .required('Username Required'),
                    password: Yup.string()
                    .max(20, 'Must be 20 characters or less')
                    .required('Password Required'),
                })}
                onSubmit={handleSubmit} 
                >
                <Form className='FormContainer'>
                    <div>
                        <label htmlFor="username">Username</label>
                        <Field id="username" name="username" type="text" placeholder="Username"/>
                        <span className='ErrorMessageContainer'><ErrorMessage name="username"/></span>
                    </div>
                    <div>
                        <label htmlFor="password">Password</label>
                        <Field id="password" label="password" name="password" type="password" placeholder="Password" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="password" /></span>
                    </div>      

                    <button type="submit">Submit</button>
                    {loading && <Loading />}
                </Form>
            </Formik>
            <Link to={`/signup`} data-testid="user-signup-link">
                <p>Don't have an account?</p>
            </Link>
        </div>
    );
};

export default LoginForm;