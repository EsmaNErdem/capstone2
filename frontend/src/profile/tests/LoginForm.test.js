import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom';
import LoginForm from "../LoginForm"

test("renders without crashing", async () => {
    render(
        <MemoryRouter >
            <LoginForm />
        </MemoryRouter>
    ); 
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(
        <MemoryRouter >
            <LoginForm />
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test("LoginForm renders correctly", () => {
    const { getByLabelText, getByText, getByTestId } = render(
        <MemoryRouter >
            <LoginForm />
        </MemoryRouter>
    );
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const submitButton = getByText("Submit");
  
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
    
    const profileLink = getByTestId('user-signup-link');
    expect(profileLink.getAttribute('href')).toBe('/signup');
  });
  
  test("Submitting the form calls login function", async () => {
    const mockLogin = jest.fn().mockResolvedValue({ success: true }); 
    const { getByLabelText, getByText } = render(
          <MemoryRouter >
              <LoginForm login={mockLogin} />
          </MemoryRouter>
      );
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const submitButton = getByText("Submit");
  
    fireEvent.change(usernameInput, { target: { value: "username" } });
    fireEvent.change(passwordInput, { target: { value: 123456789 } });
    fireEvent.click(submitButton);
  
    await waitFor(() => {
        expect(mockLogin).toHaveBeenCalledWith({
          username: "username",
          password: "123456789"
        });
    });
  });
   
