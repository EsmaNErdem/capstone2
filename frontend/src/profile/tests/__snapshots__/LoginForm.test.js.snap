// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="MuiTypography-root MuiTypography-h2 css-vqgaav-MuiTypography-root"
    >
      Login
    </h2>
    <form
      action="#"
      class="FormContainer"
    >
      <div>
        <label
          for="username"
        >
          Username
        </label>
        <input
          id="username"
          name="username"
          placeholder="Username"
          type="text"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          label="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
    <a
      data-testid="user-signup-link"
      href="/signup"
    >
      <p>
        Don't have an account?
      </p>
    </a>
  </div>
</DocumentFragment>
`;
